File: InstaDownload/app/src/main/java/com/vakarux/instadownload/InstagramDownloader.kt
package com.vakarux.instadownload

import okhttp3.Cookie
import okhttp3.CookieJar
import okhttp3.HttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONArray
import org.json.JSONObject
import java.math.BigInteger
import java.util.concurrent.TimeUnit
import java.util.regex.Pattern

data class MediaResult(
    val url: String,
    val isVideo: Boolean,
    val thumbnailUrl: String? = null,
) {
    val previewUrl: String? get() = thumbnailUrl ?: url.takeIf { !isVideo }
}

object InstagramDownloader {

    private val SHORTCODE_REGEX = Pattern.compile(
        "(?:instagram\\.com|instagr\\.am)/(?:reel|reels|p|tv)/([A-Za-z0-9_-]+)"
    )
    private val STORY_REGEX = Pattern.compile(
        "(?:instagram\\.com|instagr\\.am)/stories/([A-Za-z0-9._]+)/([0-9]+)"
    )
    private val PROFILE_REGEX = Pattern.compile(
        "^https?://(?:www\\.)?(?:instagram\\.com|instagr\\.am)/([A-Za-z0-9_.]+)/?(?:[?#].*)?$"
    )
    private val RESERVED_PROFILE_PATHS = setOf(
        "p", "reel", "reels", "tv", "stories", "explore", "accounts", "direct",
        "about", "developer", "legal", "privacy", "graphql", "web", "download", "emails", "topics"
    )

    private const val SHORTCODE_ALPHABET =
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_"

    private data class StoryRequest(val username: String, val mediaId: String)

    private val cookieStore = mutableMapOf<String, MutableList<Cookie>>()
    private val cookieJar = object : CookieJar {
        override fun saveFromResponse(url: HttpUrl, cookies: List<Cookie>) {
            cookieStore.getOrPut(url.host) { mutableListOf() }.apply {
                removeAll { c -> cookies.any { it.name == c.name } }
                addAll(cookies)
            }
        }
        override fun loadForRequest(url: HttpUrl): List<Cookie> =
            cookieStore[url.host] ?: emptyList()
    }

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .followRedirects(true)
        .cookieJar(cookieJar)
        .build()

    private val DESKTOP_UA = "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 " +
            "(KHTML, like Gecko) Chrome/142.0.0.0 Safari/537.36"
    private val MOBILE_UA = "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) " +
            "AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1"

    fun getMediaItems(
        postUrl: String,
        session: IgSession? = null,
        quality: DownloadQuality = DownloadQuality.BEST
    ): List<MediaResult> {
        extractStory(postUrl)?.let { story ->
            val s = session ?: throw UnsupportedOperationException(
                "Stories are login-only. Tap Log in with Instagram, then try again."
            )
            return tryMediaInfoApi(story.mediaId, s, quality)
        }

        val shortcode = extractShortcode(postUrl) ?: run {
            extractProfileUsername(postUrl)?.let { username ->
                return listOf(fetchProfilePicture(username))
            }
            throw IllegalArgumentException("Invalid Instagram URL: $postUrl")
        }

        val postPageError: String
        try {
            return tryPostPage(shortcode, quality)
        } catch (e: Exception) {
            postPageError = e.message ?: e.javaClass.simpleName
        }

        if (session != null) {
            try {
                return tryMediaInfoApi(shortcodeToMediaId(shortcode), session, quality)
            } catch (e: Exception) {
                throw Exception(
                    "Could not fetch this post.\n\n" +
                        "Post page: $postPageError\n" +
                        "Logged-in API: ${e.message}"
                )
            }
        }

        throw Exception(
            "Could not fetch this post — it may be private, age-restricted, or deleted.\n" +
                "Log in from the app to download content only visible to your account.\n\n" +
                "Post page: $postPageError"
        )
    }

    private fun tryMediaInfoApi(
        mediaId: String, session: IgSession, quality: DownloadQuality
    ): List<MediaResult> {
        val cookie = listOfNotNull(
            "sessionid=${session.sessionId}",
            session.csrfToken?.let { "csrftoken=$it" },
            session.userId?.let { "ds_user_id=$it" },
        ).joinToString("; ")

        val resp = client.newCall(
            Request.Builder()
                .url("https://www.instagram.com/api/v1/media/$mediaId/info/")
                .header("User-Agent", DESKTOP_UA)
                .header("X-IG-App-ID", "936619743392459")
                .header("Cookie", cookie)
                .get().build()
        ).execute()

        val body = resp.body?.string().orEmpty()
        if (body.trimStart().startsWith('<'))
            throw Exception("Media info HTTP ${resp.code}: session rejected — log in again")

        val json = runCatching { JSONObject(body) }.getOrNull()
            ?: throw Exception("Media info HTTP ${resp.code}: bad JSON — ${body.take(150)}")
        val item = json.optJSONArray("items")?.optJSONObject(0)
            ?: throw Exception("Media info HTTP ${resp.code}: ${json.optString("message").ifBlank { "no media returned" }}")

        item.optJSONArray("carousel_media")?.let { slides ->
            val items = (0 until slides.length()).mapNotNull {
                slides.optJSONObject(it)?.let { node -> extractNode(node, quality) }
            }
            if (items.isNotEmpty()) return items
        }
        return extractNode(item, quality)?.let { listOf(it) }
            ?: throw Exception("Media info: no downloadable media")
    }

    private fun extractNode(item: JSONObject, quality: DownloadQuality): MediaResult? {
        val images = item.optJSONObject("image_versions2")?.optJSONArray("candidates")
        val imageIndex = if (quality == DownloadQuality.DATA_SAVER) (images?.length() ?: 1) - 1 else 0
        val poster = images?.optJSONObject(imageIndex.coerceAtLeast(0))
            ?.optString("url")?.takeIf { it.isNotBlank() }
        val videos = item.optJSONArray("video_versions")
        val videoIndex = if (quality == DownloadQuality.DATA_SAVER) (videos?.length() ?: 1) - 1 else 0
        videos?.optJSONObject(videoIndex.coerceAtLeast(0))
            ?.optString("url")?.takeIf { it.isNotBlank() }
            ?.let { return MediaResult(it, isVideo = true, thumbnailUrl = poster) }
        return poster?.let { MediaResult(it, isVideo = false, thumbnailUrl = it) }
    }

    private fun fetchProfilePicture(username: String): MediaResult {
        val response = client.newCall(
            Request.Builder()
                .url("https://www.instagram.com/$username/")
                .header("User-Agent", "Googlebot/2.1 (+http://www.google.com/bot.html)")
                .get().build()
        ).execute()

        val html = response.body?.string()
            ?: throw Exception("Profile HTTP ${response.code}: empty body")
        if (!response.isSuccessful) throw Exception("Profile HTTP ${response.code}")

        val picUrl = Regex("""<meta property="og:image" content="([^"]+)"""")
            .find(html)?.groupValues?.get(1)?.replace("&amp;", "&")
            ?: throw Exception("Could not find a profile picture for @$username — the account may not exist")

        return MediaResult(picUrl, isVideo = false)
    }

    private fun extractProfileUsername(url: String): String? {
        val m = PROFILE_REGEX.matcher(url.trim())
        return if (m.matches()) m.group(1)?.takeUnless { it.lowercase() in RESERVED_PROFILE_PATHS } else null
    }

    fun isProfileUrl(url: String): Boolean = extractProfileUsername(url) != null

    private fun tryPostPage(shortcode: String, quality: DownloadQuality): List<MediaResult> {
        val response = client.newCall(
            Request.Builder()
                .url("https://www.instagram.com/p/$shortcode/")
                .header("User-Agent", "Googlebot/2.1 (+http://www.google.com/bot.html)")
                .get().build()
        ).execute()

        val html = response.body?.string()
            ?: throw Exception("Post HTTP ${response.code}: empty body")
        if (!response.isSuccessful) throw Exception("Post HTTP ${response.code}")

        val expectedMediaId = shortcodeToMediaId(shortcode)
        Regex("""<script\b[^>]*\bdata-sjs[^>]*>(\{.+?\})</script>""", RegexOption.DOT_MATCHES_ALL)
            .findAll(html)
            .mapNotNull { runCatching { JSONObject(it.groupValues[1]) }.getOrNull() }
            .mapNotNull { findPublicProduct(it, expectedMediaId) }
            .map { extractProductMedia(it, quality) }
            .firstOrNull { it.isNotEmpty() }
            ?.let { return it }
        throw Exception("Post HTTP ${response.code}: no public media found")
    }

    private fun findPublicProduct(value: Any?, expectedMediaId: String): JSONObject? {
        when (value) {
            is JSONObject -> {
                value.optJSONObject("if_not_gated_logged_out")?.let {
                    if (it.optString("pk") == expectedMediaId || it.optString("id") == expectedMediaId)
                        return it
                }
                if ((value.optString("pk") == expectedMediaId || value.optString("id") == expectedMediaId) &&
                    (value.has("video_versions") || value.has("carousel_media") || value.has("image_versions2")))
                    return value

                val keys = value.keys()
                while (keys.hasNext()) {
                    findPublicProduct(value.opt(keys.next()), expectedMediaId)?.let { return it }
                }
            }
            is JSONArray -> for (i in 0 until value.length()) {
                findPublicProduct(value.opt(i), expectedMediaId)?.let { return it }
            }
        }
        return null
    }

    private fun extractProductMedia(product: JSONObject, quality: DownloadQuality): List<MediaResult> {
        product.optJSONArray("carousel_media")?.let { carousel ->
            return (0 until carousel.length()).mapNotNull { i ->
                carousel.optJSONObject(i)?.let { extractNode(it, quality) }
            }
        }
        return listOfNotNull(extractNode(product, quality))
    }

    private fun mediaRequest(url: String) = Request.Builder()
        .url(url)
        .header("User-Agent", MOBILE_UA)
        .header("Referer", "https://www.instagram.com/")
        .get().build()

    fun downloadToStream(url: String, out: java.io.OutputStream) {
        val response = client.newCall(mediaRequest(url)).execute()
        if (!response.isSuccessful) throw Exception("Download HTTP ${response.code}")
        response.body?.byteStream()?.copyTo(out)
            ?: throw Exception("Empty download body")
    }

    fun fetchBytes(url: String): ByteArray {
        val response = client.newCall(mediaRequest(url)).execute()
        if (!response.isSuccessful) throw Exception("Preview HTTP ${response.code}")
        return response.body?.bytes() ?: throw Exception("Empty preview body")
    }

    private fun extractShortcode(url: String): String? {
        val m = SHORTCODE_REGEX.matcher(url)
        return if (m.find()) m.group(1)!!.take(11) else null
    }

    private fun shortcodeToMediaId(shortcode: String): String =
        shortcode.fold(BigInteger.ZERO) { id, c ->
            id * BigInteger.valueOf(64) + BigInteger.valueOf(SHORTCODE_ALPHABET.indexOf(c).toLong())
        }.toString()

    private fun extractStory(url: String): StoryRequest? {
        val m = STORY_REGEX.matcher(url)
        return if (m.find()) StoryRequest(m.group(1)!!, m.group(2)!!) else null
    }
}

File: InstaDownloadiOS/InstaDownload/Core/InstagramDownloader.swift
import Foundation

struct MediaItem: Identifiable, Hashable, Sendable {
    let url: URL
    let isVideo: Bool
    let thumbnailURL: URL?

    var id: String { url.absoluteString }

    var previewURL: URL? { thumbnailURL ?? (isVideo ? nil : url) }
}

enum DownloadError: LocalizedError {

    case invalidURL(String)

    case unsupported(String)

    case failed(String)

    var errorDescription: String? {
        switch self {
        case .invalidURL(let url): "Invalid Instagram URL: \(url)"
        case .unsupported(let message): message
        case .failed(let message): message
        }
    }
}

private struct RE {
    private let regex: NSRegularExpression

    init(_ pattern: String, options: NSRegularExpression.Options = []) {
        guard let regex = try? NSRegularExpression(pattern: pattern, options: options) else {
            preconditionFailure("Invalid regular expression literal: \(pattern)")
        }
        self.regex = regex
    }

    func firstGroup(_ index: Int = 1, in text: String) -> String? {
        let text = text as NSString
        let range = NSRange(location: 0, length: text.length)
        guard let match = regex.firstMatch(in: text as String, range: range),
              index < match.numberOfRanges
        else { return nil }
        let group = match.range(at: index)
        return group.location == NSNotFound ? nil : text.substring(with: group)
    }

    func allGroups(_ index: Int = 1, in text: String) -> [String] {
        let text = text as NSString
        let range = NSRange(location: 0, length: text.length)
        return regex.matches(in: text as String, range: range).compactMap { match in
            guard index < match.numberOfRanges else { return nil }
            let group = match.range(at: index)
            return group.location == NSNotFound ? nil : text.substring(with: group)
        }
    }

    func matches(_ text: String) -> Bool {
        let text = text as NSString
        return regex.firstMatch(in: text as String, range: NSRange(location: 0, length: text.length)) != nil
    }
}

enum InstagramDownloader {

    private struct StoryRequest {
        let username: String
        let mediaID: String
    }

    private static let session: URLSession = {
        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = 30
        configuration.timeoutIntervalForResource = 60
        configuration.httpShouldSetCookies = true
        configuration.httpCookieAcceptPolicy = .always
        configuration.httpCookieStorage = HTTPCookieStorage.shared
        configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
        return URLSession(configuration: configuration)
    }()

    private static let desktopUA = "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 "
        + "(KHTML, like Gecko) Chrome/142.0.0.0 Safari/537.36"
    private static let mobileUA = "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) "
        + "AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1"

    private static let shortcodeRE = RE(#"(?:instagram\.com|instagr\.am)/(?:reel|reels|p|tv)/([A-Za-z0-9_-]+)"#)
    private static let storyRE = RE(#"(?:instagram\.com|instagr\.am)/stories/([A-Za-z0-9._]+)/([0-9]+)"#)
    private static let validPostRE = RE(#"^https?://(www\.)?(instagram\.com|instagr\.am)/(p|reel|tv)/[A-Za-z0-9_-]+"#)
    private static let storyURLRE = RE(#"^https?://(www\.)?instagram\.com/stories/[A-Za-z0-9._]+/?.*$"#)

    private static let videoURLRE = RE(#"\\"video_url\\":\\"(https:(?:(?!\\").)*)"#)
    private static let displayURLRE = RE(#"\\"display_url\\":\\"(https:(?:(?!\\").)*)"#)
    private static let isVideoRE = RE(#"\\"is_video\\":true"#)
    private static let embeddedImageRE = RE(#"<img[^>]*class="EmbeddedMediaImage"[^>]*src="(https:[^"]*)""#)
    private static let ogImageRE = RE(#"<meta property="og:image" content="([^"]+)""#)
    private static let sjsScriptRE = RE(
        #"<script\b[^>]*\bdata-sjs[^>]*>(\{.+?\})</script>"#,
        options: [.dotMatchesLineSeparators]
    )
    private static let percentEscapeRE = RE(#"%[0-9A-Fa-f]{2}"#)

    static func isValidPostURL(_ url: String) -> Bool { validPostRE.matches(url) }

    static func isStoryURL(_ url: String) -> Bool { storyURLRE.matches(url) }

    static func mediaItems(for postURL: String) async throws -> [MediaItem] {
        if let story = extractStory(from: postURL) {
            return try await publicStory(story)
        }

        guard let shortcode = extractShortcode(from: postURL) else {
            throw DownloadError.invalidURL(postURL)
        }

        let embedFailure: String
        do {
            return try await tryEmbedPage(shortcode: shortcode)
        } catch let error as DownloadError {
            embedFailure = error.errorDescription ?? "no media found"
        } catch {
            embedFailure = error.localizedDescription
        }

        do {
            return try await tryPostPage(shortcode: shortcode)
        } catch {
            let postFailure = (error as? DownloadError)?.errorDescription ?? error.localizedDescription
            throw DownloadError.failed("""
                Could not fetch this post. It may be private, age-restricted, or deleted. \
                This build only downloads public content. Use the login build for private posts and stories.

                Embed: \(embedFailure)
                Post page: \(postFailure)
                """)
        }
    }

    private static func publicStory(_ story: StoryRequest) async throws -> [MediaItem] {
        let userID = try await fetchPublicUserID(username: story.username)
        let reels = try await fetchPublicReelsMedia(
            userID: userID,
            mediaID: story.mediaID,
            username: story.username
        )
        let items = extractStoryMedia(reels, userID: userID, mediaID: story.mediaID)
        if !items.isEmpty { return items }

        let reelCount = (reels["reels"] as? [String: Any])?.count ?? 0
        throw DownloadError.unsupported("""
            Instagram did not expose this story through public anonymous endpoints. \
            Resolved @\(story.username) to user id \(userID), but reels_media returned \
            \(reelCount) reel(s). Normal stories require a logged-in session or may have expired.
            """)
    }

    private static func fetchPublicUserID(username: String) async throws -> String {
        let encoded = username.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? username
        let url = URL(string: "https://www.instagram.com/api/v1/users/web_profile_info/?username=\(encoded)")!
        let (data, response) = try await get(url, headers: [
            "User-Agent": desktopUA,
            "Accept": "*/*",
            "Accept-Language": "en-US,en;q=0.9",
            "Referer": "https://www.instagram.com/\(username)/",
            "X-IG-App-ID": "936619743392459",
            "X-ASBD-ID": "129477",
            "X-Requested-With": "XMLHttpRequest",
        ])

        guard (200..<300).contains(response.statusCode) else {
            let body = decodeText(data) ?? ""
            throw DownloadError.failed("Profile HTTP \(response.statusCode): \(body.prefix(200))")
        }

        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let user = (json["data"] as? [String: Any])?["user"] as? [String: Any]
        else {
            throw DownloadError.failed("Profile response did not include user data")
        }

        if user["is_private"] as? Bool == true {
            throw DownloadError.unsupported("@\(username) is private")
        }

        guard let id = string(user["id"]) else {
            throw DownloadError.failed("Profile response did not include user id")
        }
        return id
    }

    private static func fetchPublicReelsMedia(
        userID: String,
        mediaID: String,
        username: String
    ) async throws -> [String: Any] {
        let url = URL(
            string: "https://www.instagram.com/api/v1/feed/reels_media/?reel_ids=\(userID)&media_id=\(mediaID)"
        )!
        let (data, response) = try await get(url, headers: [
            "User-Agent": desktopUA,
            "Accept": "*/*",
            "Accept-Language": "en-US,en;q=0.9",
            "Referer": "https://www.instagram.com/stories/\(username)/\(mediaID)/",
            "X-IG-App-ID": "936619743392459",
            "X-ASBD-ID": "129477",
            "X-Requested-With": "XMLHttpRequest",
        ])

        let body = decodeText(data) ?? ""
        if body.trimmingCharacters(in: .whitespacesAndNewlines).hasPrefix("<") {
            throw DownloadError.failed("Story HTTP \(response.statusCode): got HTML instead of JSON")
        }
        guard (200..<300).contains(response.statusCode) else {
            throw DownloadError.failed("Story HTTP \(response.statusCode): \(body.prefix(200))")
        }
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw DownloadError.failed("Story HTTP \(response.statusCode): response was not a JSON object")
        }
        return json
    }

    private static func extractStoryMedia(
        _ json: [String: Any],
        userID: String,
        mediaID: String
    ) -> [MediaItem] {
        guard let reels = json["reels"] as? [String: Any] else { return [] }

        guard let reel = (reels[userID] as? [String: Any])
            ?? reels.values.compactMap({ $0 as? [String: Any] }).first
        else { return [] }

        guard let items = reel["items"] as? [[String: Any]] else { return [] }
        for item in items {
            let id = string(item["id"]) ?? ""
            let pk = string(item["pk"]) ?? ""
            if id == mediaID || id.hasPrefix("\(mediaID)_") || pk == mediaID {
                return [mediaItem(from: item)].compactMap { $0 }
            }
        }
        return []
    }

    private static func mediaItem(from item: [String: Any]) -> MediaItem? {
        let poster = ((item["image_versions2"] as? [String: Any])?["candidates"] as? [[String: Any]])?
            .first.flatMap { string($0["url"]) }
            ?? string(item["display_url"])

        if let videoURL = (item["video_versions"] as? [[String: Any]])?
            .first.flatMap({ string($0["url"]) }),
           let url = URL(string: videoURL) {
            return MediaItem(
                url: url,
                isVideo: true,
                thumbnailURL: poster.flatMap { URL(string: $0) }
            )
        }

        if let poster, let url = URL(string: poster) {
            return MediaItem(url: url, isVideo: false, thumbnailURL: url)
        }

        return nil
    }

    private static func embedRefusal(in html: String) -> String? {
        if html.contains("Please wait a few minutes before you try again") {
            return "Instagram is rate-limiting this device. Wait a few minutes and try again."
        }

        if html.contains(#""contextJSON":null"#) {
            return "Instagram would not serve this post to a logged-out client. "
                + "That usually means it is age-restricted or login-gated, but it can also be "
                + "private, deleted, or region-blocked; the embed page doesn't say which. "
                + "This build is public-only; the login build can fetch it."
        }

        if !html.contains(#""contextJSON":""#) && html.contains("/accounts/login/") {
            return "Instagram served a login wall. This build only downloads public content."
        }

        return nil
    }

    private static func tryEmbedPage(shortcode: String) async throws -> [MediaItem] {
        let url = URL(string: "https://www.instagram.com/p/\(shortcode)/embed/captioned/")!
        let (data, response) = try await get(url, headers: [
            "User-Agent": mobileUA,
            "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            "Accept-Language": "en-US,en;q=0.5",
            "Referer": "https://www.instagram.com/",
        ])

        guard let html = decodeText(data) else {
            throw DownloadError.failed("Embed HTTP \(response.statusCode): empty body")
        }
        guard (200..<300).contains(response.statusCode) else {
            throw DownloadError.failed("Embed HTTP \(response.statusCode): \(html.prefix(120))")
        }

        if let refusal = embedRefusal(in: html) {
            throw DownloadError.unsupported(refusal)
        }

        if let video = videoURLRE.firstGroup(in: html) {
            let poster = displayURLRE.firstGroup(in: html).map(unescape)
            guard let url = URL(string: unescape(video)) else {
                throw DownloadError.failed("Embed HTTP \(response.statusCode): malformed video URL")
            }
            return [MediaItem(url: url, isVideo: true, thumbnailURL: poster.flatMap { URL(string: $0) })]
        }

        if isVideoRE.matches(html) {
            throw DownloadError.failed(
                "Embed HTTP \(response.statusCode): is_video=true but video_url missing from embed response"
            )
        }

        let displayImages = uniqueImages(from: displayURLRE.allGroups(in: html))
        if !displayImages.isEmpty { return displayImages }

        let embeddedImages = uniqueImages(from: embeddedImageRE.allGroups(in: html))
        if !embeddedImages.isEmpty { return embeddedImages }

        let ogImages = uniqueImages(from: ogImageRE.allGroups(in: html))
        if !ogImages.isEmpty { return ogImages }

        throw DownloadError.failed(
            "Embed HTTP \(response.statusCode): no media URL found (\(html.count) chars)"
        )
    }

    private static func uniqueImages(from rawURLs: [String]) -> [MediaItem] {
        var seen = Set<String>()
        return rawURLs.compactMap { raw in
            let unescaped = unescape(raw)
            guard seen.insert(unescaped).inserted, let url = URL(string: unescaped) else { return nil }
            return MediaItem(url: url, isVideo: false, thumbnailURL: nil)
        }
    }

    private static func tryPostPage(shortcode: String) async throws -> [MediaItem] {
        let url = URL(string: "https://www.instagram.com/p/\(shortcode)/")!
        let (data, response) = try await get(url, headers: [
            "User-Agent": "Googlebot/2.1 (+http://www.google.com/bot.html)",
        ])

        guard let html = decodeText(data) else {
            throw DownloadError.failed("Post HTTP \(response.statusCode): empty body")
        }
        guard (200..<300).contains(response.statusCode) else {
            throw DownloadError.failed("Post HTTP \(response.statusCode)")
        }

        let expectedMediaID = try mediaID(fromShortcode: shortcode)
        for payload in sjsScriptRE.allGroups(in: html) {
            guard let json = try? JSONSerialization.jsonObject(with: Data(payload.utf8)),
                  let product = findPublicProduct(json, expectedMediaID: expectedMediaID)
            else { continue }
            let items = extractProductMedia(product)
            if !items.isEmpty { return items }
        }

        throw DownloadError.failed("Post HTTP \(response.statusCode): no public media found")
    }

    private static func findPublicProduct(_ value: Any, expectedMediaID: String) -> [String: Any]? {
        if let object = value as? [String: Any] {
            if let ungated = object["if_not_gated_logged_out"] as? [String: Any],
               string(ungated["pk"]) == expectedMediaID || string(ungated["id"]) == expectedMediaID {
                return ungated
            }

            let matchesID = string(object["pk"]) == expectedMediaID || string(object["id"]) == expectedMediaID
            let hasMedia = object["video_versions"] != nil
                || object["carousel_media"] != nil
                || object["image_versions2"] != nil
            if matchesID && hasMedia { return object }

            for nested in object.values {
                if let found = findPublicProduct(nested, expectedMediaID: expectedMediaID) { return found }
            }
        } else if let array = value as? [Any] {
            for nested in array {
                if let found = findPublicProduct(nested, expectedMediaID: expectedMediaID) { return found }
            }
        }
        return nil
    }

    private static func extractProductMedia(_ product: [String: Any]) -> [MediaItem] {
        if let carousel = product["carousel_media"] as? [[String: Any]] {
            return carousel.compactMap { mediaItem(from: $0) }
        }
        return [mediaItem(from: product)].compactMap { $0 }
    }

    private static func mediaID(fromShortcode shortcode: String) throws -> String {
        let alphabet = Array("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_")
        var id: Int64 = 0
        for character in shortcode {
            guard let digit = alphabet.firstIndex(of: character) else {
                throw DownloadError.invalidURL("Invalid Instagram shortcode")
            }
            let (scaled, scaleOverflow) = id.multipliedReportingOverflow(by: 64)
            let (next, addOverflow) = scaled.addingReportingOverflow(Int64(digit))
            guard !scaleOverflow, !addOverflow else {
                throw DownloadError.invalidURL("Invalid Instagram shortcode")
            }
            id = next
        }
        return String(id)
    }

    private static func mediaRequest(_ url: URL) -> URLRequest {
        var request = URLRequest(url: url)
        request.setValue(mobileUA, forHTTPHeaderField: "User-Agent")
        request.setValue("https://www.instagram.com/", forHTTPHeaderField: "Referer")
        return request
    }

    static func data(for url: URL) async throws -> Data {
        let (data, response) = try await session.data(for: mediaRequest(url))
        guard let http = response as? HTTPURLResponse else {
            throw DownloadError.failed("Download failed: no HTTP response")
        }
        guard (200..<300).contains(http.statusCode) else {
            throw DownloadError.failed("Download HTTP \(http.statusCode)")
        }
        return data
    }

    private static func get(_ url: URL, headers: [String: String]) async throws -> (Data, HTTPURLResponse) {
        var request = URLRequest(url: url)
        for (field, value) in headers {
            request.setValue(value, forHTTPHeaderField: field)
        }
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw DownloadError.failed("No HTTP response from \(url.host() ?? url.absoluteString)")
        }
        return (data, http)
    }

    private static func decodeText(_ data: Data) -> String? {
        String(data: data, encoding: .utf8) ?? String(data: data, encoding: .isoLatin1)
    }

    private static func string(_ value: Any?) -> String? {
        if let string = value as? String { return string.isEmpty ? nil : string }
        if let number = value as? NSNumber { return number.stringValue }
        return nil
    }

    private static let escapedAmpersand = "\\" + "u0026"

    private static func unescape(_ raw: String) -> String {
        var value = raw
            .replacingOccurrences(of: #"\\\/"#, with: "/")
            .replacingOccurrences(of: escapedAmpersand, with: "&")
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&quot;", with: "\"")
            .replacingOccurrences(of: "&#039;", with: "'")
            .replacingOccurrences(of: "&lt;", with: "<")
            .replacingOccurrences(of: "&gt;", with: ">")

        if percentEscapeRE.matches(value), let decoded = value.removingPercentEncoding {
            value = decoded
        }
        return value
    }

    private static func extractShortcode(from url: String) -> String? {
        shortcodeRE.firstGroup(in: url).map { String($0.prefix(11)) }
    }

    private static func extractStory(from url: String) -> StoryRequest? {
        guard let username = storyRE.firstGroup(1, in: url),
              let mediaID = storyRE.firstGroup(2, in: url)
        else { return nil }
        return StoryRequest(username: username, mediaID: mediaID)
    }
}
